linked_list/linked_list.py

```diff
@@ -13,67 +13,126 @@ def __init__(self):
 
     # returns the value in the first node
     # returns None if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def get_first(self):
-        pass
+        if self.head == None:
+            return None
+        return self.head.value
 
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(self, value):
-        pass
+        new_head = Node(value, self.head)
+        self.head = new_head
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        cur = self.head
+        while cur != None:
+            if cur.value == value:
+                return True
+            cur = cur.next
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        cur = self.head
+        i = 0
+        while cur != None:
+            cur = cur.next
+            i += 1
+        return i
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        cur = self.head
+        for i in range(index):
+            if cur == None:
+                return None
+            cur = cur.next
+        return cur.value
+
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        if self.head == None:
+            return None
+        cur = self.head
+        prev = None
+        while cur != None:
+            prev = cur
+            cur = cur.next
+        return prev.value

```

Comment:
another way we can do this to save a variable is starting the while loop at `while cur.next != None`:

```suggestion
        cur = self.head
        while cur.next != None:
            cur = cur.next
        return cur.value
```